proposals/0013-alignas.md

```diff
@@ -0,0 +1,278 @@
+
+## Motivation
+
+Some GPUs can optimize 16-byte memory accesses for buffer loads and stores.
+Unfortunately, in many instances, IHV compilers must assume 4-byte alignments
+for structured buffer element accesses.
+
+In the current specification, UAV buffer root views may be aligned to 4 bytes,
+whereas as descriptors in descriptor tables must be aligned to 256 bytes.
+Therefore, when an application chooses to use root views, the IHV compiler must
+assume the worst-case alignment.
+For some GPUs, this will disable vectorized loads and stores from memory that
+require 16-byte aligned addresses.
+
+Under consideration for future specifications, placed resource alignment
+requirements may be tightened to as much as 1-byte alignment.
+This change would further justify maintaining minimum alignment requirements
+and additional explicit application hints.
+
+Currently, these limitations can only be optimized using runtime monitoring and
+bookkeeping of root view addresses, followed by background thread
+re-compilation and re-caching of shaders.
+
+## Proposed solution
+
+The `alignas` specifier may be specified on the declaration of a
+structure or the declaration of a structure member, when used by
+`[RW]StructuredBuffer` declarations, and templated `ByteAddressBuffer` loads
+and stores.
+
+### Syntax
+
+The following excepts from the
+[C++ reference manual](https://en.cppreference.com/w/cpp/language/alignas)
+would apply to HLSL.
+
+> **alignas**( _expression_ )  
+> **alignas**( _type-id_ )  
+> 1. _expression_ must be an integral constant expression that evaluates to
+> zero, or to a valid value for an alignment or extended alignment.
+> 2. Equivalent to `alignas(alignof(type-id))`.
+
+> The `alignas` specifier may be applied to:
+> - the declaration or definition of a class;
+> - the declaration of a non-bitfield class data member;
+
+> The object or the type declared by such a declaration will have its
+> alignment requirement equal to the strictest (largest) non-zero expression
+> of all `alignas` specifiers used in the declaration, unless it would weaken
+> the natural alignment of the type.
+
+> If the strictest (largest) `alignas` on a declaration is weaker than the
+> alignment it would have without any `alignas` specifiers (that is, weaker
+> than its natural alignment or weaker than `alignas` on another declaration
+> of the same object or type), the program is ill-formed.
+
+> Invalid non-zero alignments, such as `alignas(3)` are ill-formed.
+
+> Valid non-zero alignments that are weaker than another `alignas` on the same
+> declaration are ignored.
+
+> `alignas(0)` is always ignored.
+
+### Example Usage
+
+```c++
+///////////////////////////////////////////////////////////////////////////////
+// common.h
+// every object of type Foo will be aligned to 16-bytes
+struct alignas(16) Foo
+{ 
+    float3 bar; 
+    alignas(16) uint baz; // 16-byte aligned member
+}; 
+static_assert(sizeof(Foo) == 32);
```

Comment:
RawBufferLoad is a SPIR-V only thing, and this proposal also targets that.

would be nice if `static_assert` got implemented in the SPIR-V backend (currently only works in DXIL).